### PR TITLE
[netcore] System.Runtime.InteropServices test fixes.

### PR DESCRIPTION
**Pull request by vargaz**

### CODE REVIEW

File: mcs/class/System.Private.CoreLib/System.Runtime.InteropServices/Marshal.cs [hunk 1]
```diff
@@ -42,17 +42,22 @@ partial class Marshal
 		[MethodImplAttribute (MethodImplOptions.InternalCall)]
 		public extern static void StructureToPtr (object structure, IntPtr ptr, bool fDeleteOld);
 
-		internal static IntPtr AllocBSTR (int length)
-		{
-			throw new NotImplementedException ();
-		}
+		[MethodImplAttribute (MethodImplOptions.InternalCall)]
+		public extern static int GetArrayElementSize (Type type);
```
Comment (EgorBo) [inline mcs/class/System.Private.CoreLib/System.Runtime.InteropServices/Marshal.cs]: Does .NET Core have this public API? can't find it.

Comment (vargaz) [reply mcs/class/System.Private.CoreLib/System.Runtime.InteropServices/Marshal.cs]: Fixed.

**Comment by vargaz** [comment]
@monojenkins build failed

**Comment by vargaz** [comment]
@monojenkins build failed

File: mono/metadata/marshal.c [hunk 2]
```diff
@@ -5187,6 +5199,36 @@ ptr_to_structure (gconstpointer src, MonoObjectHandle dst, MonoError *error)
...
+	MONO_CHECK_ARG_NULL_NAMED (src, "ptr",);
+	MONO_CHECK_ARG_NULL_HANDLE_NAMED (dst, "structure",);
+
+	t = m_class_get_byval_arg (mono_handle_class (dst));
+	if (!allow_vtypes && MONO_TYPE_ISSTRUCT (t)) {
+		mono_error_set_argument (error, "structure", "The structure must not be a value class.");
+		return;
+	}
+
+#ifdef ENABLE_NETCORE
```
Comment (marek-safar) [inline mono/metadata/marshal.c]: Why the double #ifdef?

File: mono/metadata/marshal.c [hunk 3]
```diff
@@ -5187,6 +5199,36 @@ ptr_to_structure (gconstpointer src, MonoObjectHandle dst, MonoError *error)
...
 
+#ifdef ENABLE_NETCORE
+
+void
+ves_icall_System_Runtime_InteropServices_Marshal_PtrToStructureInternal (gconstpointer src, MonoObjectHandle dst, MonoBoolean allow_vtypes, MonoError *error)
+{
+	MonoType *t;
+	MonoClass *klass;
+
+	MONO_CHECK_ARG_NULL_NAMED (src, "ptr",);
```
Comment (marek-safar) [inline mono/metadata/marshal.c]: Wouldn't be easier to do the arguments check in managed code?

File: mono/metadata/object-internals.h [hunk 4]
```diff
@@ -38,6 +38,15 @@
 	} 								\
 } while (0)
 
+#define MONO_CHECK_ARG_NULL_NAMED(arg, argname, retval) do {	\
+	if (G_UNLIKELY (!(arg)))				\
+	{							\
+		ERROR_DECL (error);				\
+		mono_error_set_argument_null (error, (argname), "");	\
+		mono_error_set_pending_exception (error);	\
+		return retval;					\
```
Comment (jaykrell) [inline mono/metadata/object-internals.h]: Just a nit: You can remove `ERROR_DECL` and `mono_error_set_pending_exception`, in the uses so far. Perhaps not in future uses.

**Comment by vargaz** [comment]
@monojenkins build failed

**Comment by vargaz** [comment]
@monojenkins build failed

**Comment by vargaz** [comment]
@monojenkins build failed

**Comment by vargaz** [comment]
@monojenkins build failed

**Comment by vargaz** [comment]
@monojenkins build failed

**Pull request merged**